### DISCUSSION TITLE
How to use Psi after building from source (specifically for components which require source build)

### DISCUSSION THREAD

**Discussion by trigaten** [microsoft/psi#167] [Q&A]
![Psi](https://user-images.githubusercontent.com/42948753/121594476-1c1c4d80-ca0b-11eb-94e6-941962dc2b23.PNG)
After building Psi from source, I have the folders/files in the image. 

If I make a new project in the Psi solution, how do I use Psi components? I tried `using Microsoft.Psi` in the default Program.cs of a basic Console App, but the compiler can't find the Microsoft.Psi package. Are there any tutorial videos that go through developing after building from source?

**Comment by sandrist** [comment]
You'll need to first add `Microsoft.Psi` (and any other class libraries of components you might need) as "Project References" to your project. In Visual Studio, this can be done by right-clicking on your new project, selecting "Add -> Project Reference..."

![image](https://user-images.githubusercontent.com/7461717/121616352-9a7eec80-ca17-11eb-9f85-5c88fce95bff.png)

For example, once you've added the `Microsoft.Psi` project reference, your new project's .csproj file will look something like the following:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi\Microsoft.Psi.csproj" />
  </ItemGroup>

</Project>
```

**Reply by trigaten** [reply]
Thank you. I followed your instructions and added these items to the .csproj:
```
<ItemGroup>
    <ProjectReference Include="..\Sources\Imaging\Microsoft.Psi.Imaging\Microsoft.Psi.Imaging.csproj" />
    <ProjectReference Include="..\Sources\RealSense\Microsoft.Psi.RealSense.Windows.x64\Microsoft.Psi.RealSense.Windows.x64.csproj" />
    <ProjectReference Include="..\Sources\RealSense\Microsoft.Psi.RealSense_Interop.Windows.x64\Microsoft.Psi.RealSense_Interop.Windows.x64.vcxproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi.Interop\Microsoft.Psi.Interop.csproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi.Windows\Microsoft.Psi.Windows.csproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi\Microsoft.Psi.csproj" />
  </ItemGroup>
```

I ran a simple code snippet:

```
using System;
using Microsoft.Psi;
using Microsoft.Psi.RealSense.Windows;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var p = Pipeline.Create();
            RealSenseSensor r = new(p);
            //r.ColorImage.Do(t => Console.WriteLine(t));
            p.RunAsync();
            Console.ReadKey();
            p.Dispose();
        }
    }
}
```

but I get the following error:

![BFE](https://user-images.githubusercontent.com/42948753/121726723-efbf0a80-cab8-11eb-8369-19e0e0f8e028.PNG)

I have the correct reference in the .csproj (I believe?).

Why might this error be occurring?

**Comment by chitsaw** [comment]
This error usually indicates a mismatch in the project's platform target settings. The RealSense component is x64 only, so your project's Platform target also needs to be x64, as shown below:
![image](https://user-images.githubusercontent.com/32501077/121740980-bb485000-cab2-11eb-9787-73093dcca4c1.png)

**Reply by trigaten** [reply]
I made that change, but am getting the same error. I am taking a look at [https://stackoverflow.com/questions/10275106/badimageformatexception-x64-issue](https://stackoverflow.com/questions/10275106/badimageformatexception-x64-issue), but am not sure if their answers will apply. Do you have any other ideas?

**Reply by xiangzhi** [reply]
Another possibility is that you are using `netcoreapp3.1`(.NET Core) framework?  Psi uses the `.windows` prefix to specify certain package to be windows only, which uses `net472` (.NET Framework). That framework only works on Windows.  Can you check your project's csproj and see what kind of project it is? For example:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi\Microsoft.Psi.csproj" />
  </ItemGroup>

</Project>
```
If you look at `<TargetFramework>netcoreapp3.1</TargetFramework>` it is a .NET core project. If its .NET framework, it should be `net472`

I believe the Intel Real Sense component is Windows only? So your program has to be a .NET framework program.

**Reply by trigaten** [reply]
Thank you. I was using `netcoreapp3.1`. I have changed to `net472`, which got me past that error. Now I am getting the following:

System.IO.FileNotFoundException: 'Could not load file or assembly 'Microsoft.Psi.RealSense_Interop.Windows.x64.dll' or one of its dependencies. The specified module could not be found.'

I am quite certain that I built from source correctly (including the RealSense support). Do you have any ideas as to why VS can't find this file?

Here is my current .csproj:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net472</TargetFramework>
    <Platforms>AnyCPU;x64</Platforms>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
    <PlatformTarget>x64</PlatformTarget>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\Sources\Imaging\Microsoft.Psi.Imaging\Microsoft.Psi.Imaging.csproj" />
    <ProjectReference Include="..\Sources\RealSense\Microsoft.Psi.RealSense.Windows.x64\Microsoft.Psi.RealSense.Windows.x64.csproj" />
    <ProjectReference Include="..\Sources\RealSense\Microsoft.Psi.RealSense_Interop.Windows.x64\Microsoft.Psi.RealSense_Interop.Windows.x64.vcxproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi.Interop\Microsoft.Psi.Interop.csproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi.Windows\Microsoft.Psi.Windows.csproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi\Microsoft.Psi.csproj" />
  </ItemGroup>

</Project>
```

**Reply by xiangzhi** [reply]
It seemed like the dependencies didn't build correctly. Can you try building just the depend files and see what happen? To do that, hover your cursor above the `Microsoft.Psi.RealSense_Interop.Windows.x64`, right click and select build. It's likely to fail but it will tell us what is wrong.

**Reply by trigaten** [reply]
I built that, but it did not fail.

**Reply by xiangzhi** [reply]
I don't have any other good suggestions. The Psi team probably have better insight into it. 

If you are open to trying other random debugging tricks, here's some things I would try/validate:
1. Delete the bin/obj folder in your project directory and rebuild. See if it works after that.
2. Check if the `dll` file is in same folder as your generated bin file. The path should be like `PROJECT/bin/debug/net472/`. If its not there, try copying it from the other folders.

**Reply by chitsaw** [reply]
Could you also try modifying your .csproj file as follows (replacing/deleting `AnyCPU`):

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net472</TargetFramework>
    <Platforms>x64</Platforms>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
    <PlatformTarget>x64</PlatformTarget>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\Sources\Imaging\Microsoft.Psi.Imaging\Microsoft.Psi.Imaging.csproj" />
    <ProjectReference Include="..\Sources\RealSense\Microsoft.Psi.RealSense.Windows.x64\Microsoft.Psi.RealSense.Windows.x64.csproj" />
    <ProjectReference Include="..\Sources\RealSense\Microsoft.Psi.RealSense_Interop.Windows.x64\Microsoft.Psi.RealSense_Interop.Windows.x64.vcxproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi.Interop\Microsoft.Psi.Interop.csproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi.Windows\Microsoft.Psi.Windows.csproj" />
    <ProjectReference Include="..\Sources\Runtime\Microsoft.Psi\Microsoft.Psi.csproj" />
  </ItemGroup>

</Project>
```

**Reply by trigaten** [reply]
I did this, but the same issue occurs.

**Reply by trigaten** [reply]
> I don't have any other good suggestions. The Psi team probably have better insight into it.
> 
> If you are open to trying other random debugging tricks, here's some things I would try/validate:
> 
> 1. Delete the bin/obj folder in your project directory and rebuild. See if it works after that.
> 2. Check if the `dll` file is in same folder as your generated bin file. The path should be like `PROJECT/bin/debug/net472/`. If its not there, try copying it from the other folders.

1: Tried, but no luck.
2: There is no debug folder in PROJECT/bin. Is that an issue?
